### DISCUSSION TITLE
Record the time of each key press and release during typing

### DISCUSSION THREAD

**Discussion by zhenghuang17** [jspsych/jsPsych#2966] [Q&A]
Hi, I would like to start an experiment which asks participants to type things. Is it possible to record each time stamp when participant makes a key press or release, as well as showing what is typing in on the screen?

Thanks for help!

**Comment by ChristopheBossens** [comment]
I think the easiest way to achieve this is to use the plugin-survey-text plugin (so that you show what a participant is typing on the screen). You can adapt the plugin to also include the time stamps of key presses as data:

Using jsPsych 7.2.3, you can adapt the file plugin-survey-text.js as follows:

On line 187 (before the comment that says //backup in case autofocus doesn't work), add the following:

```
          // Add callback for keydown events on input
          var key_presses = []
          document.getElementById("input-0").addEventListener("keydown", (e) => {
            key_presses.push({
                'keycode' : e['code'],
                'rt'      : performance.now() - startTime
            })
          })
```

This creates a data structure in which key presses are stored. It also adds an event listener to the first input field that listens to key presses. Note that with the way it is currently implemented, it will only work for the first input (the survey plugin allows you to use multiple input fields). It also only works in the specific case that your input field only has 1 row. 

To add the data to the output of the plugin, scroll a little bit down to the comment that mentions // save data. Here you can add the key_presses:

```
// save data
var trialdata = {
   rt: response_time,
   response: question_data,
   keypresses: key_presses
};
```

Using the plugin as follows will allow you to keep track of the key_presses:

```
 var input_trial = {
   type: jsPsychSurveyText,
   questions: [
      {prompt: 'How old are you?'}
    ]
}
```

**Reply by zhenghuang17** [reply]
Hi Christophe, thanks so much for this detailed response! In my case, I would probably need to record the time of all the key presses (and releases if possible) for every word being typed in. And participants will see every character as they type the word, but not from previous words. Is it possible to record the event for each character? Thanks again!

**Reply by ChristopheBossens** [reply]
Hi zorahuang17,

It is not completely clear to me how what you describe differs from what the code already does. You can present different words using different `jsPsychSurveyText `trials, and with the adjustment it will save all the key presses.

Could you create some example code of your own of how different trials will look? From what I understand you should be able to do that using some of the jsPsych plugins and we can then see how we can fit in the time of all the key presses.

**Reply by zhenghuang17** [reply]
Hi Christophe, I've been written in MATLAB and the code is a bit long to be attached, but I've written a pseudo code here. What I'm a bit confused about your answer is whether I can record the time of press and release of each character in every word.

```
for i = 1: 100 % start trial, supposing 100 words are generated for this trial
  if ProbeTrial
      present probe % probes are presented on a trial designed to have probes, they are delivered at a random timepoint
  end
  if keyDown % keep track when each word starts
      while 1 % record every character until the word ends (ENTER pressed)
          if ENTERKeyPressed
              record press & release time of the ENTER key;
              break;
          elseif OtherKeyPressed
              record the key press identity & time;
          elseif OtherKeyReleased
              record the key release identitiy & time;
          end
      end
  end
end
```
